README.md
# Laravel Health

[![Latest Version on Packagist](https://img.shields.io/packagist/v/mindtwo/laravel-health.svg?style=flat-square)](https://packagist.org/packages/mindtwo/laravel-health)
[![GitHub Tests Action Status](https://img.shields.io/github/actions/workflow/status/mindtwo/laravel-health/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/mindtwo/laravel-health/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/actions/workflow/status/mindtwo/laravel-health/fix-php-code-style-issues.yml?branch=main&label=code%20style&style=flat-square)](https://github.com/mindtwo/laravel-health/actions?query=workflow%3A"Fix+PHP+code+style+issues"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/mindtwo/laravel-health.svg?style=flat-square)](https://packagist.org/packages/mindtwo/laravel-health)

This is where your description should go. Limit it to a paragraph or two. Consider adding a small example.

## Installation

You can install the package via composer:

```bash
composer require mindtwo/laravel-health
```

You can publish the config file with:

```bash
php artisan vendor:publish --tag="laravel-health-config"
```

This is the contents of the published config file:

```php
return [
];
```

## Usage

```php
Coming soon...
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [mindtwo GmbH](https://github.com/mindtwo)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-health` will be documented in this file.


tests/Pest.php
<?php

use Mindtwo\LaravelHealth\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/TestCase.php
<?php

namespace Mindtwo\LaravelHealth\Tests;

use Mindtwo\LaravelHealth\Checks\DummyCheck;
use Mindtwo\LaravelHealth\LaravelHealthServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected $systemReport;

    protected function setUp(): void
    {
        parent::setUp();
        $this->systemReport = new DummyCheck();
    }

    protected function getPackageProviders($app): array
    {
        return [
            LaravelHealthServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app): void
    {
        //$this->config()->set('database.default', 'testing');
    }

    public function testSystemName(): void
    {
        $this->assertEquals('Laravel', $this->systemReport->getSystemName());
    }

    public function testSystemVersion(): void
    {
        $this->assertEquals('10.30.0', $this->systemReport->getSystemVersion());
    }

    public function testPhpVersion(): void
    {
        $this->assertEquals(phpversion(), $this->systemReport->getPhpVersion());
    }

    public function testOsName(): void
    {
        $this->assertEquals(php_uname('s'), $this->systemReport->getOsName());
    }

    public function testOsVersion(): void
    {
        $this->assertEquals(php_uname('r'), $this->systemReport->getOsVersion());
    }

    public function testWebserverName(): void
    {
        $this->assertEquals($_SERVER['SERVER_SOFTWARE'], $this->systemReport->getWebservername());
    }

    public function testWebserverVersion(): void
    {
        $this->assertEquals('', $this->systemReport->getWebserverVersion());
    }

    public function testDatabaseName(): void
    {
        $this->assertEquals(config('database.default'), $this->systemReport->getDatabaseName());
    }

    public function testDatabaseVersion(): void
    {
        $this->assertEquals(config('database.connections.'.config('database.default').'.version'), $this->systemReport->getDatabaseVersion());
    }

    public function testDependenciesName(): void
    {
        $this->assertEquals(config('dependencies_name', []), $this->systemReport->getDependenciesName());
    }

    public function testDependenciesVersion(): void
    {
        $this->assertEquals(config('dependencies_version', []), $this->systemReport->getDependenciesVersion());
    }

    public function testAdditionalLastUpdate(): void
    {
        $this->assertEquals(now()->toIso8601String(), $this->systemReport->getAdditionalLastUpdate());
    }
}


src/Interfaces/DummyInterface.php
<?php

namespace Mindtwo\LaravelHealth\Interfaces;

interface DummyInterface
{
    /**
     * Get the name of the main system or frame (e.q. Laravel)
     */
    public function getSystemName(): string;

    /**
     * Get the version of the main system or frame (e.q. 10.0)
     */
    public function getSystemVersion(): string;

    /**
     * Get the php version of the main system of frame (e.q. 8.0)
     */
    public function getPhpVersion(): string;

    /**
     * Get the os name of the main system or frame (e.q. Ubuntu)
     */
    public function getOsName(): string;

    /**
     * Get the os version of the main system or frame (e.q. 20.04 LTS)
     */
    public function getOsVersion(): string;

    /**
     * Get the webserver name of the main system or frame (e.q. Apache)
     */
    public function getWebserverName(): string;

    /**
     * Get the webserver version of the main system or frame (e.q. 2.4)
     */
    public function getWebserverVersion(): string;

    /**
     * Get the database name of the main system or frame (e.q. MySQL)
     *
     * @return ?string
     */
    public function getDatabaseName(): ?string;

    /**
     * Get the database version of the main system or frame (e.q. 5.7)
     *
     * @return ?string
     */
    public function getDatabaseVersion(): ?string;

    /**
     * Get the dependencies name of the main system or frame (e.q. Nova)
     *
     * @return ?array
     */
    public function getDependenciesInfos(): ?array;

    /**
     * Get the last update  of the main system or frame (e.q. 2023-10-24T18:00:00Z)
     */
    public function getAdditionalLastUpdate(): string;
}


src/Controllers/DummyController.php
<?php

namespace Mindtwo\LaravelHealth\Controllers;

use Illuminate\Http\JsonResponse;
use Illuminate\Routing\Controller;
use Illuminate\Support\Facades\Config;
use Mindtwo\LaravelHealth\Checks\DummyCheck;

class DummyController extends Controller
{
    public function show(): JsonResponse
    {
        // If the ip is not in the whitelist, return 403

        $ipWhitelist = Config::get('system-report.ip_whitelist');
        $clientIp = request()->ip();
        if (! in_array($clientIp, $ipWhitelist)) {

            return response()->json(['error' => 'Forbidden.'], 403);
        }
        //dd($ipWhitelist);
        //dd(Config::get('system-report.ip_whitelist'));

        //$output = shell_exec('mysql -V');

        // dd(DB::select('SELECT version()'));

        $systemReport = new DummyCheck();

        return response()->json($systemReport->toArray());

    }
}


src/Checks/DummyCheck.php
<?php

namespace Mindtwo\LaravelHealth\Checks;

use Illuminate\Support\Facades\App;
use Illuminate\Support\Facades\DB;
use Mindtwo\LaravelHealth\Interfaces\DummyInterface;
use Spatie\Health\Checks\Result;
use Throwable;

class DummyCheck implements DummyInterface
{
    public function run(): Result
    {
        $result = Result::make()->meta([
            'test' => 'success',
        ]);

        return $result->ok();
    }

    /**
     * Get the name of the main system or frame (e.q. Laravel)
     */
    public function getSystemName(): string
    {
        return 'Laravel';
    }

    /**
     * Get the version of the main system or frame (e.q. 10.0)
     */
    public function getSystemVersion(): string
    {
        return App::version();
    }

    /**
     * Get the php version of the main system of frame (e.q. 8.0)
     */
    public function getPhpVersion(): string
    {
        return phpversion();
    }

    /**
     * Get the os name of the main system or frame (e.q. Ubuntu)
     */
    public function getOsName(): string
    {
        return php_uname('s');
    }

    /**
     * Get the os version of the main system or frame (e.q. 20.04 LTS)
     */
    public function getOsVersion(): string
    {
        return php_uname('r');
    }

    /**
     * Get the webserver name of the main system or frame (e.q. Apache)
     */
    public function getWebserverName(): string
    {
        $serverSoftware = explode('/', $_SERVER['SERVER_SOFTWARE']);

        return $serverSoftware[0] ?? '';
    }

    /**
     * Get the webserver version of the main system or frame (e.q. 2.4)
     */
    public function getWebserverVersion(): string
    {
        $serverSoftware = explode('/', $_SERVER['SERVER_SOFTWARE']);

        return $serverSoftware[1] ?? '';
    }

    /**
     * Get the database name of the main system or frame (e.q. MySQL)
     *
     * @return ?string
     */
    public function getDatabaseName(): ?string
    {
        return config('database.default');
    }

    /**
     * Get the database version of the main system or frame (e.q. 5.7)
     *
     * @return ?string
     */
    public function getDatabaseVersion(): ?string
    {
        // $error = 'The system is not healthy';
        try {
            $version = DB::select('SELECT version() AS version');

            return $version[0]->version;
        } catch (Throwable $error) {
            return null;
        }
    }

    /**
     * Get the dependencies name of the main system or frame (e.q. Nova)
     *
     * @return ?array
     */
    public function getDependenciesInfos(): ?array
    {
        return null;
    }

    /**
     * Get the dependencies version of the main system or frame (e.q. 4.1.0)
     */
    public function getDependenciesVersion(): string
    {
        return config('dependencies_version', []);
    }

    /**
     * Get the last update  of the main system or frame (e.q. 2023-10-24T18:00:00Z)
     */
    public function getAdditionalLastUpdate(): string
    {
        return now()->toIso8601String();
    }

    public function toArray(): array
    {
        return [
            'system' => [
                'type' => $this->getSystemName(),
                'version' => $this->getSystemVersion(),
            ],
            'os' => [
                'name' => $this->getOsName(),
                'version' => $this->getOsVersion(),
            ],
            'php' => [
                'version' => $this->getPhpVersion(),
            ],
            'webserver' => [
                'name' => $this->getWebserverName(),
                'version' => $this->getWebserverVersion(),
            ],
            'database' => [
                'name' => $this->getDatabaseName(),
                'version' => $this->getDatabaseVersion(),
            ],
            'dependencies' => [
                'name' => $this->getDependenciesInfos(),
                'version' => $this->getDependenciesInfos(),
            ],
            'additional' => [
                'lastUpdated' => $this->getAdditionalLastUpdate(),
            ],
        ];
    }
}
